src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala

```diff
@@ -58,7 +60,7 @@ class TaskReplaceActor(
   val oldActiveInstances = oldInstances.filter(_.state.goal == Goal.Running)
 
   // All instances to kill as set for quick lookup
-  private[this] var oldInstanceIds: SortedSet[Id] = oldActiveInstances.map(_.instanceId).to[SortedSet]
+  private[this] var oldActiveInstanceIds: SortedSet[Id] = oldActiveInstances.map(_.instanceId).to[SortedSet]
```

Comment:
why do we need these old instance ids anyway? Why can't we just work with runspec versions? And kill/decommission everything that has different runspec version than the one we're trying to go to?

Reply:
this would simplify stuff nicely because we can just kill everything that does not belong to our runspec version :) because we know it can't be the new instance id, because we are starting new instances exclusively with the new version...

Reply:
That's a good point. We use it to count down the old instances and then check the set in `checkFinished`. @zen-dog and I thought we should just aggregate the current state like I did in https://github.com/mesosphere/marathon/tree/karsten/state-replace-actor but in a simplified way. I would like to do that in a follow up.

Reply:
well if we do that, we don't need this because then on this line https://github.com/mesosphere/marathon/pull/6893/files#diff-799d40bf745dd8a6faf58e52993998b8R183 we could just kill the task that is not terminal and with different version and that will pretty nicely go around the need to queue any operations because the orchestration logic will be able to take care of these inconsistencies by itself

Reply:
I still don't get it. Sorry. The issue is that we might miss the instance completely. Let's say we are rolling back to two instance. The `TaskReplaceActor` sees two instances and does not know a third is queued. The actor would then just finish and never see the update. Also, I had a solution with a third `else if` branch in `// === An InstanceChanged event for the *old* instance ===`. @zen-dog and I agreed that this was more complicated that queuing the updates which we have to do anyways.